sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4

```diff
@@ -578,6 +578,21 @@ namedExpressionSeq
     : namedExpression (',' namedExpression)*
     ;
 
+transformList
```

Comment:
why is this not a general expression?

Reply:
Because in this context, only transforms are allowed. You can't partition by an arbitrary expression.

Reply:
shouldn't the parser be generic, and we just throw errors on things that are not supported? that'd give better error messages (typical technique used by more mature databases) and also reduce the number of parsing rules in the grammar file.

Reply:
We can definitely do that and you're right about being able to produce better error messages.

Is it okay with you if we do this in a follow-up?

This PR is blocking the addition of `TableCatalog` in PR #24246, and that PR is blocking many PRs that we can do in parallel (Select from SQL, CreateTable, CTAS, DropTable, etc.). If we are going to get this done in time, I'd like to get these changes in and work on improvements like this later. I'd rather have a functional v2 implementation than friendly error messages right now.

Reply:
It's a good idea to make the parser rule general and produce a better error message. However this is an existing problem in our parser, and it's also a non-trivial work for this case. We need to catch `UnresolvedFunction` and valid the arguments, the arguments can only be attributes or `ExtractValues` or literal. I'm ok to do it in a followup.

Reply:
I've opened an issue for this follow-up: https://issues.apache.org/jira/browse/SPARK-27386